### ISSUE TITLE
Swift package template dependencies

### ISSUE THREAD

**Issue by mattmassicotte** [tree-sitter/tree-sitter#3392]
### Problem

Hello!

I was attempting to use the Swift bindings for the OCaml parser, and found something very surprising. That parser has a dependency on SwiftTreeSitter. I didn't realize that this had all been created by the CLI templates, but the maintainer of that parser pointed me here.

https://github.com/tree-sitter/tree-sitter-ocaml/issues/92

I've manually added SPM (Swift) support to 20-30 parsers myself. And I'm the author of SwiftTreeSitter - I'm happy for people to use it! But, I think it is unnecessarily restritive to force it as a dependency for all parser clients. We can definitely acheive the testing goals without it.

And, now that I'm here, I see a few improvements that can be made to the package structure that seem worthwhile.

Are you all open to making changes here?

### Steps to reproduce

-

### Expected behavior

-

### Tree-sitter version (tree-sitter --version)

-

### Operating system/version

-

**Comment by amaanq** [comment]
>  I didn't realize that this had all been created by the CLI templates, but the maintainer of that parser pointed me here.

Yeah this was just added a few days ago to add testing support for most of the bindings

> Are you all open to making changes here?

Totally! PR welcome

**Comment by mattmassicotte** [comment]
> It had not. OCaml had Swift tests _before_ those were properly (?) added to the template.

Ah!

I think these tests are great! They are just testing more than is necessary to prove a functioning implemenation. They can be both simpler and dependency-free.


> See also [tree-sitter-grammars/tree-sitter-xml#17](https://github.com/tree-sitter-grammars/tree-sitter-xml/pull/17).

Hmm, yes. Repos that define more than one parser usually require special-casing. I haven't used that one myself, so I'm not 100% sure how to best handle it. Two seperate modules does seem reasonable, given that its likely only one could be needed.

**Comment by amaanq** [comment]
Since there hasn't been any follow-up in a while, I'll close this out. I'm also assuming it's "solved" because the template only marks `SwiftTreeSitter` as a test dependency only, but if that's problematic let me know (the core grammars have been updated w/ this for months now, so I'm assuming it's fine).

(Side note, would you be willing to upstream SwiftTreeSitter to the org? It's a great binding library that would be nice to have "upstream", which would help its visibility, usage, and credibility :^), but no worries if not)

**Comment by mattmassicotte** [comment]
I wish, but unfortunately, SPM has no concept of "test-only" dependencies. So yes, this is still a problem.  I'm sorry I haven't looked into it yet!

Wow upstreaming it is something I have never thought about! I guess the only downside I can think of is the project has undergone a number of pretty major changes, and usually they are quite unilateral on my part. What would the process for change look like?

**Comment by amaanq** [comment]
Ah okay, sorry about that. No worries on not looking into it yet, it would be nice if we could have a parser test that the parser can be loaded independent of SwiftTreeSitter - I'll reopen this.

As for upstreaming, the process would include probably adding me as a collaborator with write access (or maybe admin, GitHub doesn't make this stuff easy), then I can transfer it to the upstream org. You'd still retain full admin rights for that repo, and the changes for development wouldn't be too jarring I'd think. In general, bindings follow a few "guidelines" (really just things I do myself, but maybe it could be documented for contributors somewhere):

- The bindings should follow updates to the library on every release, if applicable, meaning adding the appropriate code for any features added upstream

- The bindings should follow language conventions as best as it can - for example, in C, getting the root node involves calling `ts_tree_root_node`, but in Rust, this is just a method on the `Tree` struct, `Tree::root_node`, so calls look like `tree.root_node()`. But in Python, `root_node` is a property of `Tree`, since "computing" this is pretty cheap, and it's more Python-like.

- Typically, breakage to public APIs should be avoided as best as possible, and there shouldn't be "extra" features in the bindings that ordinarily would be something the user of the library should make (such as a visitor API)

Of course, there's no pressure to upstream the bindings, but I think it'd be really nice :slightly_smiling_face:

**Comment by mattmassicotte** [comment]
Let's do it! https://github.com/ChimeHQ/SwiftTreeSitter/issues/32

**Comment by amaanq** [comment]
cool!!
